daffodil-core/src/main/scala/org/apache/daffodil/core/dpath/Expression.scala

```diff
@@ -2829,13 +2834,23 @@ sealed abstract class LengthExprBase(
   ) {
 
   protected final def leafReferencedElements = {
-    val arg = args(0).asInstanceOf[PathExpression]
-    val steps = arg.steps
+    val path = args(0) match {
+      case arg: PathExpression => arg
+      case _ => {
+        // $COVERAGE-OFF$
+        Assert.invariantFailed("NodeInfo.Exists, but arg was not a PathExpression.")
+        // $COVERAGE-ON$
+      }
+    }
+    // test_lengthPatternBinaryPatternLimit in section 12 shows that the length can
+    // be evaluated when it is an array of hex binary values
+    if (!path.isSelf && path.isPathToOneWholeArray)
```

Comment:
Does the `path.isSelf` condition want to be checked inside `isPathToOneWholeArray`. A self path can never be a path to a whole array, so I imagine we would need this same check anywhere `isPathToOneWholeArray`, or we could just include it in he logic?

Reply:
I moved the `path.isSelf` check to be inside `isPathToOneWholeArray` and that seems to work.

Reply:
This is quite clean now. I like it.